src/linker/pipeline.py

```diff
@@ -87,3 +45,92 @@ def _validate_implementations(self) -> Dict:
+                self.implementations[idx - 1], results_dir
+            )
+            return [str(previous_output_dir / "result.parquet")]
+
+    def get_output_dir(self, implementation: Implementation, results_dir: Path) -> Path:
+        idx = self.implementation_indices[implementation.name]
+        if idx == len(self.implementations) - 1:
+            return results_dir
+
+        return results_dir / "intermediate" / self.get_step_id(implementation)
+
+    def get_diagnostics_dir(self, implementation: Implementation, results_dir: Path) -> Path:
+        return results_dir / "diagnostics" / self.get_step_id(implementation)
+
+    def build_snakefile(self, results_dir: Path) -> Path:
+        self.write_imports(results_dir)
+        self.write_target_rules(results_dir)
+        for implementation in self.implementations:
+            self.write_implementation_rules(implementation, results_dir)
+        return results_dir / "Snakefile"
+
+    def write_imports(self, results_dir) -> None:
+        snakefile = results_dir / "Snakefile"
+        with open(snakefile, "a") as f:
+            f.write("import linker.utilities.validation_utils as validation_utils")
+
+    def write_target_rules(self, results_dir) -> None:
+        final_output = [str(results_dir / "result.parquet")]
+        validator_file = str(results_dir / "input_validations" / "final_validator")
+        target_rule = TargetRule(target_files=final_output, validation=validator_file)
+        final_validation = ValidationRule(
+            name="results",
+            input=final_output,
+            output=validator_file,
+            validator=validate_dummy_file,
+        )
+        target_rule.write_to_snakefile(results_dir)
+        final_validation.write_to_snakefile(results_dir)
+
+    def write_implementation_rules(self, implementation, results_dir) -> None:
+        input_files = self.get_input_files(implementation, results_dir)
+        output_files = [
+            str(self.get_output_dir(implementation, results_dir) / "result.parquet"),
+        ]
+        diagnostics_dir = self.get_diagnostics_dir(implementation, results_dir)
+        diagnostics_dir.mkdir(parents=True, exist_ok=True)
+        validation_file = str(
+            results_dir / "input_validations" / implementation.validation_filename
+        )
+        validation_rule = ValidationRule(
+            name=implementation.name,
+            input=input_files,
+            output=validation_file,
+            validator=implementation.step.validate_file,
+        )
+        implementation_rule = ImplementedRule(
+            name=implementation.name,
+            execution_input=input_files,
+            validation=validation_file,
+            output=output_files,
+            envvars=implementation.environment_variables,
+            diagnostics_dir=str(diagnostics_dir),
+            script_cmd=implementation.script_cmd,
+        )
+        validation_rule.write_to_snakefile(results_dir)
```

Comment:
The order here matters very much, right? Maybe a quick note about that and why.

Reply:
renamed class ValidationRule to InputValidation rule and added a comment!